### DISCUSSION TITLE
Supabase auth?

### DISCUSSION THREAD

**Discussion by magick93** [vendurehq/vendure#2488] [Q&A]
Hello

I'm currently building on Supabase. Supabase uses postgres and also ships with its own auth. 

The supabase auth is _not_ an OIDC IDP. So I dont think it can be integrated as another 'social' style login. 

I'm wondering if it would be possible, using a plugin, to replace the implementation of the default auth and use Supabase auth?

**Comment by michaelbromley** [comment]
Hi,
It should be possible - we have a pretty flexible mechanism for supporting external auth providers: https://docs.vendure.io/guides/core-concepts/auth/#external-authentication

**Reply by magick93** [reply]
Thanks @michaelbromley 

I'm studying some of the existing auth strategies, such as `packages/dev-server/test-plugins/keycloak-auth/keycloak-authentication-strategy.ts`. 

A couple of questions for you:

1.  Does the `test-plugins` mean they are recommended for testing?
2. Does a 3rd party auth provider result in, and allow vendure to use, the 3rd party provided jwt or venture signed jwt?

The reason I'm asking the last question is Supabase uses postrest, and has some useful postgres functions to examine the incoming http request and check for the jwt. For example, the use of `auth.uid()` in the database. 

```sql

create policy "Users can insert their own profile."
  on profiles for insert
  with check ( auth.uid() = id );
``` 

So, I'm wondering, does the vendure expect an authenticated result from a 3rd party, and then it will issue a jwt, or can it leverage anothers jwt (which can also include authorization claims)?

**Reply by magick93** [reply]
I see JWTs have [been removed](https://github.com/vendure-ecommerce/vendure/issues/25) - so I'm wondering, perhaps it would make more sense to provide an option to _also_ return the jwt from 3rd party auths?

**Reply by magick93** [reply]
Both Supabase and Vendure have session tables. So there may be a possibility to use the supabase auth, and add a trigger that adds a row to the vendure session table. 

Sorry to keep throwing question out you. But I have a few more regarding some of the columns in the session table. 

1. `type` - what is the purpose of this column? ANSWERED - this is the type of user, eg, anon or authenticated. 
2. `token` - does this need to be of any particular format?


Thanks

**Reply by michaelbromley** [reply]
Hi,

> Does the test-plugins mean they are recommended for testing?

It means I use that dir to test out certain plugin functionality.

> token - does this need to be of any particular format?

No, not really. It's just a random string. Here's how it gets generated: https://github.com/vendure-ecommerce/vendure/blob/a817a1a672cd15b3fd86b3cea5d3dc7404dc064d/packages/core/src/service/services/session.service.ts#L321

**Reply by magick93** [reply]
Thanks @michaelbromley 

Regarding the `user` table:

```sql

CREATE TABLE vendure."user" (
	"createdAt" timestamp NOT NULL DEFAULT now(),
	"updatedAt" timestamp NOT NULL DEFAULT now(),
	"deletedAt" timestamp NULL,
	identifier varchar NOT NULL,
	verified bool NOT NULL DEFAULT false,
	"lastLogin" timestamp NULL,
	id serial4 NOT NULL,
	CONSTRAINT "PK_cace4a159ff9f2512dd42373760" PRIMARY KEY (id)
);
```

Is the `identifier` primarily intended to store email addresses? Is there a reason this is not a `unique` column?



Thanks

**Reply by michaelbromley** [reply]
Yes often it would be an email address, but sometimes other identifies are used such as phone number for OTP-based logins.

 It is not unique at the DB-level because the entity is soft-deletable (`deletedAt` column), and enforcing a `unique` constraint would prevent someone signing up with the same identifier as a deleted user.

**Reply by magick93** [reply]
Thanks @michaelbromley

**Comment by siquick** [comment]
Did you implement this @magick93? Any recommendations for approach?
